### PR TITLE
Code Refactoring for the Dumpulator Package

### PR DESCRIPTION
**Pull request by anthonyprintup**
Currently the project consists of a large codebase which at first glance isn't friendly to new developers. I've made some changes in an attempt to make the code more easily readable/understandable, and correct (in cases where bugs have been found).

Applicable scopes:
- [x] details
- [x] dumpulator
- [x] handles
- [x] memory
- [x] modules
- [x] native

### CODE REVIEW

**Comment by anthonyprintup** [comment]
As a note, I think it would be a good idea to get rid of typing imports for standard collections (refer to [this](https://docs.python.org/3/whatsnew/3.9.html#type-hinting-generics-in-standard-collections) and [PEP 585](https://peps.python.org/pep-0585/)).

File: src/dumpulator/dumpulator.py [hunk 1]
```diff
@@ -1075,7 +1079,7 @@ def map_module(self, file_data: bytes, file_path: str = "", requested_base: int
             if write:
                 protect = MemoryProtect.PAGE_READWRITE
             if execute:
-                protect = MemoryProtect(protect.value << 4)
+                protect = MemoryProtect.PAGE_EXECUTE_READ
```
Comment (mrexodia) [inline src/dumpulator/dumpulator.py:1082]: There is no `elif` here, the `<< 4` is converting from PAGE_XX to PAGE_EXECUTE_XX (which I agree warrants a comment)

image

File: src/dumpulator/dumpulator.py [hunk 2]
```diff
@@ -1123,7 +1126,7 @@ def _hook_mem(uc: Uc, access, address, size, value, dp: Dumpulator):
         return True
 
     fetch_accesses = [UC_MEM_FETCH, UC_MEM_FETCH_PROT, UC_MEM_FETCH_UNMAPPED]
-    if access == UC_MEM_FETCH_UNMAPPED and address >= FORCE_KILL_ADDR - 0x10 and address <= FORCE_KILL_ADDR + 0x10 and dp.kill_me is not None:
+    if access == UC_MEM_FETCH_UNMAPPED and FORCE_KILL_ADDR - 0x10 <= address <= FORCE_KILL_ADDR + 0x10 and dp.kill_me is not None:
```
Comment (mrexodia) [inline src/dumpulator/dumpulator.py:1129]: I don't find this more readable

File: src/dumpulator/dumpulator.py [hunk 3]
```diff
@@ -1178,7 +1181,7 @@ def _hook_mem(uc: Uc, access, address, size, value, dp: Dumpulator):
                 assert value == dp.exception.memory_value
 
                 # Delete the code hook
-                uc.hook_del(int(dp.exception.code_hook_h))
+                uc.hook_del(dp.exception.code_hook_h.value)
```
Comment (mrexodia) [inline src/dumpulator/dumpulator.py:1184]: Shouldn't `exception.code_hook_h` just be an `int` as unicorn returns it, to not have to do these gymnastics?

File: src/dumpulator/memory.py [hunk 4]
```diff
@@ -104,28 +107,28 @@ def read(self, addr: int, size: int) -> bytearray:
...
-    def __init__(self, page_manager: PageManager, minimum=0x10000, maximum=0x7fffffff0000, granularity=0x10000):
-        self._page_manager = page_manager
-        self._minimum = minimum
-        self._maximum = maximum
-        self._granularity = granularity
-        self._regions: List[MemoryRegion] = []
-        self._committed: Dict[int, MemoryRegion] = {}
+    _page_manager: PageManager
+    _minimum: int = 0x000000010000
+    _maximum: int = 0x7fffffff0000
```
Comment (mrexodia) [inline src/dumpulator/memory.py:128]: Either this should be 64-bits wide or as it was before, this is confusing

File: src/dumpulator/memory.py [hunk 5]
```diff
@@ -60,7 +63,7 @@ def __lt__(self, other: Any):
 
     def __contains__(self, other: object) -> bool:
         if isinstance(other, int):
-            return other >= self.start and other < self.end
+            return self.start <= other < self.end
```
Comment (mrexodia) [inline src/dumpulator/memory.py:66]: Ditto, less readable

File: src/dumpulator/memory.py [hunk 6]
```diff
@@ -160,7 +165,7 @@ def align_allocation(self, addr: int) -> int:
         return (addr + mask) & ~mask
 
     def find_free(self, size: int, allocation_align=True) -> Optional[int]:
-        assert size > 0 and self.align_page(size) == size
+        assert 0 < size == self.align_page(size)
```
Comment (mrexodia) [inline src/dumpulator/memory.py:168]: Also less readable

File: src/dumpulator/modules.py [hunk 7]
```diff
@@ -1,32 +1,39 @@
...
-        self.pe = pe
-        path = path.replace("/", "\\")
-        self.path = path
-        self.name = path.split("\\")[-1]
-        self._exports_by_address: Dict[int, int] = {}
-        self._exports_by_ordinal: Dict[int, int] = {}
-        self._exports_by_name: Dict[str, int] = {}
-        self.exports: List[ModuleExport] = []
+    pe: pefile.PE
+    path: str  # TODO(printup): use pathlib.Path
```
Comment (mrexodia) [inline src/dumpulator/modules.py:18]: I think using `pathlib.Path` might cause some confusion on lunix, no?

File: src/dumpulator/modules.py [hunk 8]
```diff
@@ -1,32 +1,39 @@
...
+    def __post_init__(self):
+        self.path = self.path.replace("/", "\\")
+        self.name = self.path.split("\\")[-1]
         self._parse_pe()
 
     def _parse_pe(self):
-        self.base: int = self.pe.OPTIONAL_HEADER.ImageBase
-        self.size: int = self.pe.OPTIONAL_HEADER.SizeOfImage
-        self.entry: int = self.base + self.pe.OPTIONAL_HEADER.AddressOfEntryPoint
+        self.base = self.pe.OPTIONAL_HEADER.ImageBase
```
Comment (mrexodia) [inline src/dumpulator/modules.py:34]: These types should stay, pelib doesn't have type hints

**Review by mrexodia — CHANGES_REQUESTED** [review]
Mostly good, appreciate the work. Just some minor things I don't agree with.

File: src/dumpulator/modules.py (hunk 7 above)
Comment (anthonyprintup) [reply src/dumpulator/modules.py:18]: The good part about `pathlib.Path` is that parsing the path (e.g. extracting the file name, extension, etc.) is built-in; therefore it should work fine on linux platforms.

File: src/dumpulator/dumpulator.py (hunk 3 above)
Comment (anthonyprintup) [reply src/dumpulator/dumpulator.py:1184]: As I understood originally the variable was supposed to use `exception.code_hook_h` (from the TODO); although I agree that adding a comment and using the value itself could also work - although less verbose.

File: src/dumpulator/dumpulator.py (hunk 1 above)
Comment (anthonyprintup) [reply src/dumpulator/dumpulator.py:1082]: Mistake on my end.

Comment (anthonyprintup) [reply src/dumpulator/dumpulator.py]: Fixed by #[6c18ee3](https://github.com/mrexodia/dumpulator/pull/47/commits/6c18ee378cc70ffad68b185cf6e50f0ea9682849)

File: src/dumpulator/memory.py (hunk 4 above)
Comment (anthonyprintup) [reply src/dumpulator/memory.py:128]: Fixed by #[6b7dc90](https://github.com/mrexodia/dumpulator/pull/47/commits/6b7dc9060c87a761a88e825e829a88ee845a79c0)

File: src/dumpulator/dumpulator.py (hunk 3 above)
Comment (anthonyprintup) [reply src/dumpulator/dumpulator.py]: Fixed by #[dcfccad](https://github.com/mrexodia/dumpulator/pull/47/commits/dcfccad8cb57b045c6fe5a3cecb60067c9920f89)

**Comment by mrexodia** [comment]
Thanks a lot!

**Pull request merged**